power_grid_model_c/power_grid_model/include/power_grid_model/component/current_sensor.hpp

```diff
@@ -137,20 +140,32 @@ template <symmetry_tag current_sensor_symmetry_> class CurrentSensor : public Ge
         return CurrentSensorCalcParam<sym_calc>{.angle_measurement_type = angle_measurement_type(),
                                                 .measurement = DecomposedComplexRandVar<sym_calc>(i_polar)};
     }
-    CurrentSensorOutput<symmetric_t> get_sym_output(ComplexValue<symmetric_t> const& i) const final {
-        return get_generic_output<symmetric_t>(i);
+    CurrentSensorOutput<symmetric_t> get_sym_output(ComplexValue<symmetric_t> const& i,
+                                                    ComplexValue<symmetric_t> const& u) const final {
+        return get_generic_output<symmetric_t>(i, u);
     }
-    CurrentSensorOutput<asymmetric_t> get_asym_output(ComplexValue<asymmetric_t> const& i) const final {
-        return get_generic_output<asymmetric_t>(i);
+    CurrentSensorOutput<asymmetric_t> get_asym_output(ComplexValue<asymmetric_t> const& i,
+                                                      ComplexValue<asymmetric_t> const& u) const final {
+        return get_generic_output<asymmetric_t>(i, u);
     }
     template <symmetry_tag sym_calc>
-    CurrentSensorOutput<sym_calc> get_generic_output(ComplexValue<sym_calc> const& i) const {
+    CurrentSensorOutput<sym_calc> get_generic_output(ComplexValue<sym_calc> const& i,
+                                                     ComplexValue<sym_calc> const& u) const {
         CurrentSensorOutput<sym_calc> output{};
         output.id = id();
-        ComplexValue<sym_calc> const i_residual{process_mean_val<sym_calc>(i_measured_ - i) * base_current_};
         output.energized = 1; // current sensor is always energized
-        output.i_residual = cabs(i_residual);
-        output.i_angle_residual = arg(i_residual);
+        auto i_output = i;
+        auto const i_calc_param = calc_param<sym_calc>();
+        auto const& angle_measurement_type = i_calc_param.angle_measurement_type;
+        auto const& i_measured_complex = i_calc_param.measurement.value();
+        if (angle_measurement_type == AngleMeasurementType::local_angle) {
+            // I_l = conj(I_g) * exp(i * u_angle)
+            // Tranform back the output angle to the local angle frame of reference
+            auto const u_angle = arg(u);
+            i_output = conj(i_output) * (cos(u_angle) + 1i * sin(u_angle));
+        }
+        output.i_residual = (cabs(i_measured_complex) - cabs(i_output)) * base_current_;
+        output.i_angle_residual = arg(i_measured_complex) - arg(i_output);
```

Comment:
I think this changes the meaning of residuals. It was complex quantity difference previously.

Reply:
indeed but it does appear to be more correct this way

Reply:
Ok I realize the way done here is correct. The local measurement is not a complex quantity. Resolving this comment.